keps/sig-storage/20200213-namespace-transfer.md

```diff
@@ -0,0 +1,265 @@
+
+Users can monitor the status of a transfer by querying for the existance of the bound target PVC or checking the `status.complete` property of `StorageTransferRequest`.
+
+`StorageTransferRequest` object:
+
+```golang
+type StorageTransferRequest struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  Spec StorageTransferRequestSpec
+  Status *StorageTransferRequestStatus
+}
+```
+
+```golang
+type StorageTransferRequestSpec struct {
+  Source *corev1.ObjectReference
+  TargetName *string
+}
+```
+
+```golang
+type StorageTransferRequestStatus struct {
+  Approved *bool
+  Complete *bool
+  Error *StorageTransferError
+}
+```
+
+`StorageTransferApproval` object:
+
+```golang
+type StorageTransferApproval struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  Spec StorageTransferApprovalSpec
+  Status *StorageTransferApprovalStatus
+}
+```
+
+```golang
+type StorageTransferApprovalSpec struct {
+  Source *corev1.TypedLocalObjectReference
+  TargetNamespace string
+}
+```
+
+```golang
+type StorageTransferApprovalStatus struct {
+  Acknowledged *bool
+  Error *StorageTransferError
+}
+```
+
+### User Stories
+
+When combined with VolumeSnapshots and PVC cloning, users are given the power to easily share/experiment with persistent data.
+
+#### Story 1
+
+The `prod` namespace contains the production database which is stored on the `database` PersistentVolumeClaim.  
+VolumeSnapshots of `database` are taken at regular intervals.  DBAs want to test a schema update script before 
+running it in `prod`.  A new PVC named `database-test` is created in `prod` from a VolumeSnapshot of `database`.  `database-test` is then transferred to the `stage` namespace where it can safely be mounted by modified.  The process is repeated until the DBAs are happy with their script.
+
+#### Story 2
+
+Someone wants to build an Infrastructure as a Service provider with Kubernetes and [KubeVirt](https://github.com/kubevirt).
+All the default virtual machine images can be stored on PersistentVolumeClaims in a single namespace called `golden-images`.  When virtual machines are created, the PVC containing the appropriate virtual machine image is cloned and transferred to the target namespace.  The virtual machine is booted from the cloned/transferred PVC.
+
+### Implementation Details/Notes/Constraints
+
+I believe the API can be implemented as CRDs and an external controller.  If so, is a KEP really required?  Would there be advantages to implementing in-tree?
+
+How do we deal with unbound PVCs (WaitForFirstConsumer)?
+
+What if a single `StorageTransferApproval` maps to multiple `StorageTransferRequests`?
+
+Should there be a way (with the API) for a user to automatically approve all transfers from a namespace?
+
```

Comment:
Do the Request and Approval objects need finalizers to prevent deletion during the critical section where the three steps comprising the actual transfer are being performed?

Reply:
Absolutely.  Not sure about including what I would consider an implementation detail here though.